openmls/src/group/tests/test_remove_operation.rs

```diff
@@ -0,0 +1,316 @@
+            .expect("Could not process unverified message.");
+
+        match bob_processed_message {
+            ProcessedMessage::StagedCommitMessage(bob_staged_commit) => {
+                bob_group
+                    .merge_staged_commit(*bob_staged_commit)
+                    .expect("An unexpected error occurred.");
+            }
+            _ => unreachable!(),
+        }
+
+        let mut charlie_group = MlsGroup::new_from_welcome(
+            backend,
+            &mls_group_config,
+            welcome,
+            Some(alice_group.export_ratchet_tree()),
+        )
+        .expect("Error creating group from Welcome");
+
+        // === Remove operation ===
+
+        let alice_kpr = *alice_group
+            .key_package_ref()
+            .expect("An unexpected error occurred.");
+        let bob_kpr = *bob_group
+            .key_package_ref()
+            .expect("An unexpected error occurred.");
+
+        // We differentiate between th two test cases here
+        let (message, _welcome) = match test_case {
+            // Alice removes Bob
+            TestCase::Remove => alice_group
+                .remove_members(backend, &[bob_kpr])
+                .expect("Could not remove members."),
+            // Bob leaves
+            TestCase::Leave => {
+                // Bob leaves the group
+                let message = bob_group
+                    .leave_group(backend)
+                    .expect("Could not leave group.");
+
+                // Alice & Charlie store the pending proposal
+                for group in [&mut alice_group, &mut charlie_group] {
+                    let unverified_message = group
+                        .parse_message(message.clone().into(), backend)
+                        .expect("Could not parse message.");
+                    let processed_message = group
+                        .process_unverified_message(unverified_message, None, backend)
+                        .expect("Could not process unverified message.");
+
+                    match processed_message {
+                        ProcessedMessage::ProposalMessage(proposal) => {
+                            group.store_pending_proposal(*proposal);
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+
+                // Alice commits to Bob's proposal
+                alice_group
+                    .commit_to_pending_proposals(backend)
+                    .expect("An unexpected error occurred.")
+            }
+        };
+
+        // === Remove operation from Alice's perspective ===
+
+        let alice_staged_commit = alice_group.pending_commit().expect("No pending commit.");
+
+        let remove_proposal = alice_staged_commit
+            .remove_proposals()
+            .next()
+            .expect("An unexpected error occurred.");
+
+        let remove_operation = RemoveOperation::new(remove_proposal, &alice_group)
```

Comment:
So the consumer inspecting the proposal has to create the `RemoveOperation` themselves? For inspection purposes, wouldn't it be more convenient for the consumer if they had all the information available immediately? Don't get me wrong, this is a great change already, I'm just wondering what we want the consumer API to look like eventually.

Reply:
As discussed, I think it's ok to just have it as an external helper for now. The current layout makes it easier to treat all remove proposals the same, which makes bulk parsing easier.
We could also add something similar for add and update proposals (although it's less needed there).